Fix mac issues

dvs opened the pull request:
This pull request introduces two checks in the MAC realignment algorithm in order to alleviate
Issue #153 ("MAC realignment algorithm damages hits").

milot-mirdita commented:
Dear Dmitry,

Thanks a lot for your contribution. Johannes and me spent some time to review your code. Sorry that we were not able to look into it earlier.

We agree that its a major issue that the MAC algorithm can return a hit outside of the Viterbi hit. However, as we understand your code, you only check if this happens on the alignment end (`hit.mac_hit_end_positions`). Shouldn't we have an equivalent procedure to check if an alignment lies outside of the beginning of the Viterbi path? Does your modified algorithm already handle this case?

We are not sure if we agree on the 90% coverage criterion. The purpose of the Viterbi is to give the best possible score/e-value estimate that the two sequences are homologous. The MAC algorithm is supposed to give us highest confidence alignment within that region. If only a short part of the alignment is sufficiently confident, then so be it.

We are open to including this option with a default 0% threshold, which will always give a maximum accuracy alignment if it was requested. Additionally, a user can also request very greedy (i.e. short, confident) alignments with a large `-mact` value (e.g., `0.7`). Could you include a code branch that would short circuit the additional MAC runs if a threshold of 0% is given?

Thanks again for your contribution.

dvs commented:
Dear Milot,

Thanks for your prompt reply (I didn't expect immediate feedback, it's summer time).

> However, as we understand your code, you only check if this happens on the alignment end (hit.mac_hit_end_positions). ...

It's not quite correct. The check is performed for the result of backtracking where both start and end coordinates in query and target spaces are known. See lines 193 and 210 where the MAC hit is considered as valid if it's segments, [i1; i2] and [j1; j2], have non-empty intersection with the Viterbi hit's segments, respectively (i.e. hits overlap in both query and target spaces). In other words, the MAC and Viterbi hits' rectangles in the DP matrix have intersection.
As for hit.mac_hit_end_positions list &ndash; its purpose is to collect alternative end positions of local MAC hits in macAlgorithm and deliver them to backtraceMAC where it could be iterated if the strongest hit turned out to be the wrong one.

> We are open to including this option with a default 0% threshold, which will always give a maximum accuracy alignment if it was requested. Additionally, a user can also request very greedy (i.e. short, confident) alignments with a large -mact value (e.g., 0.7). Could you include a code branch that would short circuit the additional MAC runs if a threshold of 0% is given?

OK. But I think there is no need to short circuit the code of additional backtracking, because even when user agrees to get shorter hits, this doesn't mean the check for the intersection is not needed (see above). We would still need to skip completely wrong hits. Therefore I've changed the default for the mac_min_length to 0.0, this would virtually mean that no checks for hit length are performed.
However this will make possible the case of the loss of significant part of the Viterbi hit after MAC realignment step, the main issue raised by my test case.
To alleviate this I would suggest to issue a warning whenever MAC hit is shorter than Viterbi hit in terms of matched columns (I added it in my recent commits). I think many users expect that MAC realignment step generally improves hits and would like to be informed otherwise.

Thank you for your attention.

Best regards,
  Dmitry

dvs commented:
P.S. I want to remind what is happening when Viterbi hit is split in two local hits during MAC realignment step: then only the strongest part is reported and the weaker part is lost.
This follows from the code's design, because backtraceMAC procedure is supposed only to improve the existing hit structure. This behavior is not desirable even when high -mact values are used, I think user would like then to get two concise local hits rather than loosing one of them.
Unfortunately it's difficult to fix this issue without major redesign of MAC algorithm's code, that's why I suggested to fallback to Viterbi hit in these cases (i.e. when -mac_min_length ratio threshold triggers).
